src/stream/src/from_proto/chain.rs

```diff
@@ -40,28 +40,31 @@ impl ExecutorBuilder for ChainExecutorBuilder {
         stream: &mut LocalStreamManagerCore,
     ) -> StreamResult<BoxedExecutor> {
         let [mview, snapshot]: [_; 2] = params.input.try_into().unwrap();
-
         // For reporting the progress.
         let progress = stream
             .context
             .register_create_mview_progress(params.actor_context.id);
 
-        // The batch query executor scans on a mapped adhoc mview table, thus we should directly use
-        // its schema.
-        let schema = snapshot.schema().clone();
-
         let output_indices = node
             .output_indices
             .iter()
             .map(|&i| i as usize)
             .collect_vec();
 
-        // For `Chain`s other than `Backfill`, there should be no extra mapping required. We can
-        // directly output the columns received from the upstream or snapshot.
-        if !matches!(node.chain_type(), ChainType::Backfill) {
-            let all_indices = (0..schema.len()).collect_vec();
+        let schema = if matches!(node.chain_type(), ChainType::Backfill) {
+            Schema::new(
+                output_indices
+                    .iter()
+                    .map(|i| snapshot.schema().fields()[*i].clone())
+                    .collect_vec(),
+            )
```

Comment:
LGTM, just adding a note here for clarity:
1. The `snapshot.schema()` is trimmed, and contains only `pk_and_output_indices` of the upstream.
2. The `output_indices` here is specially computed via `upstream_column_ids` in `stream_table_scan`, so it is accurate to use it here to project the new schema.

I guess failure happens in cases where pk columns are not part of output schema.

Reply:
Shall we add the failing case as an e2e test here as well?

Reply:
also, please notice that the stream scan's output schema must contain the upstream's "stream key" instead of table's "pk"